Fixes #20550 - support for remote DB in backup and restore

mbacovsky opened the pull request:
Backup and restore scripts were changed to access postgress DB
using the configuration from Foreman and Candlepin and not directly
as DB superuser.

In online backup we skip dump of global objects as usually there are no
privileges to restore them on the remote DB.

In offline backup we do ofline backup as before only the remote DBs
are backed up using pg_dump as in online backup because we do not have
direct access to /var/lib/pgsql/data. We assume the remote DB is always
up.

The restore was updated accordingly and can restore offline backup
including dumps of the remote dbs.

katello-service was fixed to not touch postgresql service when the DB is
remote.

mbacovsky commented:
The installer --reset is dropping the DBs and the owner usually do not have the permission to re-create them. I'll add small fix skipping the drop db for remote dbs.

Other then that I was able to backup and restore Katrello completely on remote DB.

It would be great if someone can check that the changes I had to do to the backup procedure make sense and are okay with what is expected by admins in the field.

sean797 commented on katello/katello-backup:
```diff
@@ -10,11 +10,14 @@ require 'highline/import'
 @dir = nil
 @databases = DATABASES.dup
 
+
```
Unnecessary?

sean797 commented on katello/katello-backup:
```diff
@@ -10,11 +10,14 @@ require 'highline/import'
 EXCLUDED="--exclude goferd,foreman-proxy,squid,smart_proxy_dynflow_core,qdrouterd,qpidd"
 DATABASES = ['pulp', 'pgsql', 'mongodb']
+FOREMAN_DB_CONFIG = '/etc/foreman/database.yml'
+CANDLEPIN_DB_CONFIG = '/etc/candlepin/candlepin.conf'
```
Instead of reading these 2 files, I think it would be a lot simpler to read `/etc/foreman-installer/scenarios.d/#{scenario}-answers.yaml` ?
You could get the scenario from `/etc/foreman-installer/scenarios.d/last_scenario.yaml`

mbacovsky replied:
@sean797 that's a good idea, I'll reimplement this.

sean797 commented on katello/katello-restore:
```diff
@@ -4,13 +4,17 @@ require 'socket'
 FOREMAN_PROXY_CONTENT_ONLINE_BACKUP_FILES = ['mongo_dump']
 FOREMAN_PROXY_CONTENT = "foreman-proxy-content" # keep this variable easy to change for backporting
+FOREMAN_DB_CONFIG = '/etc/foreman/database.yml'
+CANDLEPIN_DB_CONFIG = '/etc/candlepin/candlepin.conf'
```
ditto

sean797 commented on katello/katello-service:
```diff
@@ -49,7 +50,14 @@ OptionParser.new do |opts|
 end
 
+def remote_db?
+  config_file = '/etc/foreman/database.yml'
```
ditto.

sean797 commented on katello/katello-restore:
```diff
@@ -75,14 +79,47 @@ def reset_katello
+end
+
+def is_remote_db?(host)
+  host != 'localhost'
```
`127.0.0.1` & `hostname`might also be valid

sean797 commented on katello/katello-restore:
```diff
@@ -75,14 +79,47 @@ def reset_katello
+  if is_remote_db?(foreman_config[:host])
+    candlepin_config = candlepin_db_config
+    run_pg_cmd(foreman_config, 'pg_restore', "--no-privileges --clean -n public #{File.join(@dir, "foreman.dump")}")
+    run_pg_cmd(candlepin_config, 'pg_restore', "--no-privileges --clean -n public #{File.join(@dir, "candlepin.dump")}")
```
Is it possible it might be in a namespace other than `public` ?

mbacovsky replied:
@sean797 probably not, I'll remove the schema restriction

mbacovsky replied:
@sean797 I had to keep it restricted to the public schema. Otherwise it tried to clean also the procedural language and objects that the user may not be allowed to drop.

sean797 commented on katello/katello-backup:
```diff
@@ -299,19 +333,27 @@ def offline_backup(database, dir_path = nil)
-      run_cmd("tar --selinux --create --file=#{File.join(@dir, 'pgsql_data.tar')} --listed-incremental=#{File.join(@dir, '.postgres.snar')} --transform 's,^,var/lib/pgsql/data/,S' -S *")
-      puts "Done."
+    if is_remote_db?(foreman_config['host'])
+      puts "Backup of #{dir_path} is not supported for remote databases. Doing online backup instead... "
```
s/Doing online backup instead/Doing postgres online backup instead/g ? Just to be clear

sean797 reviewed (changes requested):
Thanks for doing this @mbacovsky 👍 

Since this doesn't include support for remote mongodb would you mind changing the commit message, just to be clear.
  
Couple of inline comments, generally I don't like the approach of reading '/etc/foreman/database.yml' & '/etc/candlepin/candlepin.conf'. I think the answer file would be a lot simpler to implement. I guess that does give us a dependence on the answer file, I think its fair to say everyone uses the installer to build Katello?

mbacovsky commented:
Unfortunately I'm not going to finish my updates before I leave for vacation. I'll update 1st week of September.

mbacovsky commented:
@sean797, @beav , I've updated this PR. I've extracted the common code and put it to the `/usr/share/foreman/script`. Feel free to suggest better location. I've also addressed the comments above and rebased the PR.
The `katello-restore` requires change in how the --reset feature of the installer approaches the db. The PR will follow soon.

sean797 commented:
@mbacovsky I would prefer if it could go in lib/ like https://github.com/Katello/katello-packaging/pull/524/

Overall looks good 👍  but I haven't tested it.

mbacovsky commented:
@sean797 I'll wait till the other PR is merged and update the code accordingly. `lib/` is more convenient place for the db handling code.

mbacovsky committed:
Fixes #20550 - support for remote DB in backup and restore

Backup and restore scripts were changed to access postgress DB
using the configuration from Foreman and Candlepin and not directly
as DB superuser.

In online backup we skip dump of global objects as usually there are no
privileges to restore them on the remote DB.

In offline backup we do ofline backup as before only the remote DBs
are backed up using pg_dump as in online backup because we do not have
direct access to /var/lib/pgsql/data. We assume the remote DB is always
up.

The restore was updated accordingly and can restore offline backup
including dumps of the remote dbs.

katello-service was fixed to not touch postgresql service when the DB is
remote.

mbacovsky commented:
@sean797 I've updated the patch to work with the refactored scripts. I've updated the SPEC and tested the scratch builds with rest of the packages from Katello nightly. I was able to backup and restore my Katello successfully. There were some hiccups during db:migrate and internal proxy check. It was fixed after re-running the installer. I'm not sure if that is a nightly thing or my semi-devel box upgraded from Katello 3.4.

There is one fix in progress to prevent dropping remote DB in reset hook. I'll send it later today.

sean797 commented on katello/db_config.rb:
```diff
@@ -0,0 +1,49 @@
+    end
+
+    def pg_dump_command(config, dump_file)
+      "PGPASSWORD='#{config[:password]}' pg_dump -U #{config[:username]} -h #{config[:host]} -p #{config[:port]} -Fc #{config[:database]} > #{dump_file}"
```
DRY. Use `pg_command` methods above ?

sean797 commented on katello/db_config.rb:
```diff
@@ -0,0 +1,49 @@
+          :host => answers['katello']['candlepin_db_host'] || 'localhost',
+          :port => answers['katello']['candlepin_db_port'] || '5432',
+          :ssl => answers['katello']['candlepin_ssl'] || false,
+          :ssl_verify => answers['katello']['candlepin_ssl_verify'] || false
```
These 2 SSL options don't seem to actually be used?

mbacovsky replied:
Currently not, but as it is part of the DB configuration I chose to keep it. It may become useful when building URI for JDBC like in https://github.com/Katello/katello-installer/blob/master/hooks/pre/30-upgrade.rb#L39. If it is a problem, I can remove it.

sean797 replied:
Please do, we can add them when we need them. Seem unnecessary to me :)

mbacovsky committed:
Refs #20550 - helper method for deleting all data from remote DB

mbacovsky committed:
Refs #20550 - pg_command refactoring

mbacovsky commented:
@sean797, refactored the pg_command a bit...

sean797 commented on katello/restore.rb:
```diff
@@ -82,11 +83,19 @@ def reset_katello
+        run_cmd("runuser - postgres -c 'dropdb candlepin'")
+        run_cmd("runuser - postgres -c 'psql -f #{File.join(@dir, "pg_globals.dump")} postgres 2>/dev/null'")
+        run_cmd("runuser - postgres -c 'pg_restore -C -d postgres #{File.join(@dir, "foreman.dump")}'")
+        run_cmd("runuser - postgres -c 'pg_restore -C -d postgres #{File.join(@dir, "candlepin.dump")}'")
```
nitpick. Can these work in a similar way to `backup.rb`? Making use of the methods in `DBConfig`, just for consistency. Your'll probably want to add a `pg_restore_command` method.

https://github.com/Katello/katello-packaging/pull/499/files#diff-e5c9a75a6cce52c183d1d26a2500311eR219

mbacovsky replied:
@sean797, I didn't want to touch the restore of the local DB as it is not functionally equivalent to the commands run as pg admin. However the difference is small and should affect non-standard foreman DB (custom schema changes, privileges etc.).  If you think it is safe we can remove the whole else part and use the same procedure that is used for remote DB. What do you think?

sean797 replied:
Makes sense 👍

sean797 reviewed (changes requested):
Okay, this is looking really good now! Much cleaner than before and working nicely! Thanks @mbacovsky 👍

2 small comments and then we should be good to merge 🎆

mbacovsky committed:
Refs #20550 - SSL related attrs removed

sean797 reviewed (approved):
ACK, thanks @mbacovsky 🎆 

Whoever does the merge, please squash the commits

mbacovsky commented:
@sean797 thanks for all the reviewing effort!

ehelms commented:
@cfouant can you give this a once over?

ekohl commented on katello/db_config.rb:
```diff
@@ -0,0 +1,66 @@
+    attr_reader :foreman, :candlepin
+    INSTALLER_CONFIG = '/etc/foreman-installer/scenarios.d/last_scenario.yaml'
+
+    def initialize(scenario = INSTALLER_CONFIG)
```
This is very tight coupling with the puppet modules and that is a risk if we refactor things. I'm not sure if there's a better way though and maybe we should just accept that.

ekohl reviewed (commented):
At which point do these become big enough to deserve a proper package? Is this something that at some point will migrate to foreman-maintain?

mbacovsky commented:
@ekohl there are plans to move backup and restore to the foreman-maintain and I believe it is a matter of the resources now. Is it @ehelms? 

In foreman-maintain some time ago I started the work on extending the `features` representing foreman and candlepin databases to support remote DBs. I used the service configs to obtain the db connection data so it is not necessary to couple it that tightly with puppet modules (answer files). Not sure which approach is better.

ehelms commented:
That is the general plan, although no one currently owns migrating them to foreman-maintain. But we have talked about it, and proposed it via an RFC. I don't think it would be a ton of work, more work on the testing side, and updating docs to incorporate using foreman-maintain to perform the actions versus these scripts.

ehelms commented:
These packages and code have all moved to foreman-packaging `rpm/develop` under a `katello` subfolder. Please re-open your PR against it to ensure there is no data loss between the repositories. The code in this repository will be subsequently removed -- https://github.com/Katello/katello-packaging/pull/575

johnpmitsch commented:
This functionality is being added to foreman maintain and here: https://github.com/Katello/katello-installer/pull/550
